EligibilityCheck/handler.ts

```diff
@@ -54,32 +59,62 @@ export function EligibilityCheckHandler(
 
     // If a bonus activation for that user is in progress
     // returns 403 status response
-    const maybeBonusActivationResponse = checkBonusActivationIsRunning(
-      context.bindings.processingBonusIdIn
+    const bonusProcessingFindResponse = await bonusProcessingModel.find(
+      fiscalCode,
+      fiscalCode
     );
-    if (isSome(maybeBonusActivationResponse)) {
-      return maybeBonusActivationResponse.value;
+    if (isRight(bonusProcessingFindResponse)) {
+      const maybeBonusActivationResponse = bonusProcessingFindResponse.value;
+      if (isSome(maybeBonusActivationResponse)) {
+        return ResponseErrorForbiddenNotAuthorized;
+      }
+    } else {
+      context.log.error("EligibilityCheck|ERROR|Error reading BonusProcessing");
+      return ResponseErrorInternal("Error reading the bonus processing");
     }
 
     const instanceId: InstanceId = {
       id: (fiscalCode as unknown) as NonEmptyString
     };
-
-    // If we already have a valid dsu for this user do not start the orchestrator
-    if (context.bindings.eligibilityCheck) {
-      const errorOrEligibilityCheck = EligibilityCheckSuccessEligible.decode(
-        context.bindings.eligibilityCheck
+    const hasEligibilityCheckValid = await tryCatch(
+      () => eligibilityCheckModel.find(fiscalCode, fiscalCode),
+      _ => new Error("Error reading Eligibility Check from database")
+    )
+      .chain(_ =>
+        fromEither(_).mapLeft(
+          queryError =>
+            new Error(
+              `Query error [body:${queryError.body}|code: ${queryError.code}]`
+            )
+        )
+      )
+      .chain(_ =>
+        fromEither(fromOption(new Error("Eligibility Check not found"))(_))
+      )
+      .chain(_ =>
+        fromEither(
+          EligibilityCheckSuccessEligible.decode(_).mapLeft(
+            _1 => new Error("Eligibility check in not Success Eligible")
+          )
+        )
+      )
+      .chain(
+        fromPredicate(
+          eligibilityCheck => isAfter(eligibilityCheck.validBefore, now()),
+          () => new Error("Eligibility Check Expired")
+        )
+      )
+      .fold(
+        () => false, // TODO: Logs errors?
+        () => true
+      )
+      .run();
+    if (hasEligibilityCheckValid) {
+      return ResponseSuccessRedirectToResource(
+        instanceId,
+        `/api/v1/bonus/vacanze/eligibility/${fiscalCode}`,
+        instanceId
       );
```

Comment:
What's the response in case `hasEligibilityCheckValid` is false?
 
Can we perform this check inside a `chain` or `map`?

Reply:
The renaming part of this method doesn't use `taskEither` chain. A refactoring is necessary to use only map/chain pattern

Reply:
I was wrong sorry